docs/interfaces.md

```diff
@@ -0,0 +1,571 @@
+// ECC_Enclave and TLCC_Enclave. Once a secure channel is established,
+// Session Requests to TLCC can be issued to invoke tlcc functions such
+// as getMetadata or verifyMSPId.
+
+// TODO How to integrate the the message authenticator?!
+// should we abstract it as signature here even though we may
+// use cmac once a session is established?
+
+message SessionSetup {
+        // TBD input enclave vk?
+        // TLCC Mrenclave?
+        // channel
+}
+
+message SessionSetupResponse {
+        SessionSetup setup = 1;
+        byte context       = 2;         // context that references the session
+}
+
+message SessionClose {
+        byte context = 1;
+}
+
+message SessionCloseResponse {
+}
+
+message SessionRequest {
+        byte context = 1;
+        byte nonce   = 2;
+        oneof request {
+                GetMetadataRequest metadata     = 1; 
+                validateMSPRequest validate_msp = 2;
+        }
+}
+
+message SessionRequeatResponse {
+        SessionRequest request = 1;
+        oneof response {
+                GetMetadataResponse metadata     = 1; 
+                validateMSPResponse validate_msp = 2; 
+        }
+}
+
+```
+
+
+# TLCC_Enclave
+
+## Interface:
+Defined in the TLCC EDL.
+```c++
+// TBD is this interface already sufficient to support multi-channel?
+// TODO check if a block contatains any channel information?
+
+// creates a new channel
+public int ecall_join_channel(
+        [in, size=gen_len] uint8_t *genesis, uint32_t gen_len);
+
+// pushes next block to tlcc
+public int ecall_next_block(
+        [user_check] uint8_t *block_bytes, uint32_t block_size);
+```
+
+```c++
+// this secure session functions are defined in a secure channel module
+//included from secure_channel.edl
+
+public int session_setup(
+        [in] setup SessionSetup,
+        [out] SessionSetupResponse);
+
+public int session_close(
+        [in] close SessionClose
+        [out] SessionCloseReponse);
+
+public int session_request(
+        [in] request SessionRequest,
+        [out] SessionRequestResponse;
+```
+
+```c++
+// addtionally we may use the state sealing module to persists TLCC state
+// see below the data sealing module.
+include sealed_storage.edl
+```
+
+Session requests are depatched to the following interfaces:
+```c++
+// TODO refine
+// verify state data
+public int get_state_metadata(
+        [in, string] const char *namespace,
+        [in, string] const char *key,
+        [in, size=32] uint8_t *nonce,
+        [out] sgx_cmac_128bit_tag_t *cmac);
+
+public int get_multi_state_metadata(
+        [in, string] const char *namespace,
+        [in, string] const char *comp_key,
+        [in, size=32] uint8_t *nonce,
+        [out] sgx_cmac_128bit_tag_t *cmac);
+
+// verify msp identities
+public int validate_msp_id(
+        [in, string] const char *msp_id,
+        [in, size=32] uint8_t *nonce,
+        [out] bool isValid, 
+        [out] sgx_cmac_128bit_tag_t *cmac);
```

Comment:
I guess this cmac and the cmac below would/should really be done by the session handling by ensuring request and response authenticity?  Similar all nonces can/should be done a single place as part of session handling?

Reply:
ack! nonces and some for of message authenticator are part of the session messages in `SessionRequest` and `SessionRequestResponse`

Reply:
i guess you refer to comment mentioning "signatures"?  Why signatures? We do have a shared key, so  lets just add a cmac field to both messages (or better, similar to signatures, make a sub-type of what is maced?  BTW: the two request/response types of the oneofs are not defined? But i guess they are a one-to-one mapping of the function signature ...

Reply:
ack, edited

Reply:
did you forget to push something here? Still see signatures (and no Request/Response type, although that probably could be handled just by a comment pointing to 1-1 mapping). BTW: i was referring to the sessionrequest/response, not the session setup (although, even in that case you wouldn't really would use signatures: the closest to signatures would be the DH key-exchange but that would be hidden behind the SDK API (see, e.g., https://github.com/intel/linux-sgx/blob/master/SampleCode/LocalAttestation/EnclaveInitiator/EnclaveMessageExchange.cpp for some use of the API ...)

Reply:
Can you please provide a patch fixing the session API with details based on the local attestation protocols?! :)

Reply:
will do. See below for a brain dump.  As this is really a separate module with (hopefully :-) clean interfaces, i will move the stuff to separate `.md` files ...